### PR TITLE
add typings for openapi-factory

### PR DESCRIPTION
**Pull request by runebaas**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If adding a new definition:
- [x] The package does not already provide its own types, or cannot have its `.d.ts` files generated via `--declaration`
- [x] If this is for an NPM package, match the name. If not, do not conflict with the name of an NPM package.
- [x] Create it with `dts-gen --dt`, not by basing it on an existing project.
- [x] `tslint.json` should be present, and `tsconfig.json` should have `noImplicitAny`, `noImplicitThis`, `strictNullChecks`, and `strictFunctionTypes` set to `true`.

### CODE REVIEW

**Comment by wparad** [comment]
`openapi-factory`

File: types/openapi-factory/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+        errorMiddleware?: () => any;
+    }
+}
+
+declare class OpenApi {
+    constructor(options: OpenApi.OpenApiOptions, overrideLogger?: () => void);
+
+    setAuthorizer(authorizerFunc: (req?: any) => any): void;
+    onEvent(onEventFunc: (req?: any) => any): void;
+    onSchedule(onScheduleFunc: (req?: any) => any): void;
```
Comment (wparad) [inline types/openapi-factory/index.d.ts]: There first there are async functions, should we indicate that in the type file?

File: types/openapi-factory/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    head(route: string, p0: (req?: any) => any): void;
```
Comment (wparad) [inline types/openapi-factory/index.d.ts]: is there an option to do something like:
`head(route: string, p0: (req?: object) => Promise<any>): void`

File: types/openapi-factory/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    patch(route: string, p0: (req?: any) => any): void;
+    patch(route: string, p0: object, p1: (req?: any) => any): void;
+
+    delete(route: string, p0: (req?: any) => any): void;
+    delete(route: string, p0: object, p1: (req?: any) => any): void;
+
+    options(route: string, p0: (req?: any) => any): void;
+    options(route: string, p0: object, p1: (req?: any) => any): void;
+
+    any(route: string, p0: (req?: any) => any): void;
```
Comment (wparad) [inline types/openapi-factory/index.d.ts]: * can we call `p1` (`p0` when there is no `p1`) => handler` which is a `(req?: object) => Promise<any>)`

* can we call `p0` (when there is a `p1`) options (when , and include the `Options` object here which contains the `boolean` `rawBody`?)

**Review by wparad — APPROVED** [review]
Seems good to me :+1:

File: types/openapi-factory/openapi-factory-tests.ts [hunk 4]
```diff
@@ -0,0 +1,29 @@
+import * as OpenApiFactory from 'openapi-factory';
```
Comment (amcasey) [inline types/openapi-factory/openapi-factory-tests.ts]: I think this should be `import OpenApiFactory = require('openapi-factory')` to match the CommonJS export.

File: types/openapi-factory/tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,23 @@
+{
+    "compilerOptions": {
+        "module": "commonjs",
+        "lib": [
+            "es6"
+        ],
+        "noImplicitAny": true,
+        "noImplicitThis": true,
+        "strictNullChecks": true,
+        "strictFunctionTypes": false,
```
Comment (amcasey) [inline types/openapi-factory/tsconfig.json]: Why no strictFunctionTypes?

**Review by amcasey — CHANGES_REQUESTED** [review]

File: types/openapi-factory/openapi-factory-tests.ts (hunk 4 above)
Comment (runebaas) [reply types/openapi-factory/openapi-factory-tests.ts]: Changed it

File: types/openapi-factory/tsconfig.json (hunk 5 above)
Comment (runebaas) [reply types/openapi-factory/tsconfig.json]: I was pretty sure i set that back to true, must have missed it, please excuse me

**Pull request merged**